keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+        Rule string
+        // Message customize returning message for validation error of the particular rule.
+        // +optional
+        Message string
+    }
+    ```
+
+    For validation expressions, the CEL is used. They are similar to validations functions for [Custom Resources](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#resource-use-by-validation-functions).
+    `Rule` expression should always evaluate a boolean. Token `claims` are passed to CEL expressions as a dynamic map `decls.NewMapType(decls.String, decls.Dyn)`.
+
+    You can find a snippet of validation rules below:
+
+    ```yaml
+    claimValidationRules:
+    
+    - rule: 'claims.aud == "charmander" || claims.aud == "bulbasaur"'
+      message: clients other than charmander or bulbasaur are not allowed
+    
+    - rule: 'claims.roles.exists(r, r == "kubernetes-user")'
+      message: only kubernetes-user group members can access the cluster
+    ```
+
+3. `ClaimMappings` - rules to map claims from a token to Kubernetes user attributes.
+    ```go
+    type UserAttributes struct {
+        // Username represents an option for the username attribute.
+        Username string
+        // Groups represents an option for the groups attribute. 
+        // +optional
+        Groups string
+        // UID represents an option for the uid attribute.
+        // +optional
+        UID string
+        // Extra represents an option for the extra attribute.
+        // +optional
+        Extra []ExtraMapping
+    }
+   
+    type ExtraMapping struct {
+        // Key is a CEL expression to extract extra attribute key.
+        Key string
+        // Expression is a CEL expression to extract extra attribute value.
+        Expression string
+    }
+    ```
+
+    Every field of structures above is a CEL expression. For username and uid, the expression should evaluate a string.
+    For groups, it should evaluate to an array. A map of extra attributes is composed of key/value pairs, each with their
+    own CEL expressions, both should evaluate to a string.
+
+    The example of mapping user attributes can be found below:
+
+    ```yaml
+    claimMappings:
+      username: 'claims.username + ":external-user"'
+      groups: 'claims.roles.split(",")'
+      uid: 'claims.sub'
+      extra:
+      - key: '"client_name"'
+        expression: 'claims.aud'
+    ```
+   
+    For the token with the following claims
+    ```json
+    {
+      "sub": "119abc",
+      "aud": "kubernetes",
+      "username": "jane_doe",
+      "roles": "admin,user",
+      ...
+    }
+    ```
+    our authenticator will extract the following user attributes:
+    ```yaml
+    username: jane_doe:external-user
+    groups: ["admin", "user"]
+    uid: "119abc"
+    extra:
+      client_name: kubernetes
+    ```
+4. `ClaimsFilter` - list of claim names that should be passed to CEL expressions. The assumption is that administrators
```

Comment:
can the cel expressions be introspected to see what claims are used and auto-filter rather than require double configuration?

Reply:
We discussed it with @jpbetz on Slack, but did not manage to find a better solution. I want to use the `decls.NewMapType(decls.String, decls.Dyn)` for claims, so there is no way to no for sure which fields are accessed during on the expression evaluation.

> Maksim Nabokikh
  10:25 PM
The claims struct is a map, but keys and values may vary. For now, we unmarshal a JSON document into an unstructured map and then pass it to CEL. To avoid overloading, there is a claims filter. If the user knows that only particular claims are required to operate (like name, groups, and aud in my previous example), it is possible to drop other claims and not unmarshal them.

> 10:27 PM
We can assume that users know the schema of tokens provided by their providers, but I did not manage to find a way to benefit from this knowledge.

> jpbetz
10:30 PM
Gotcha, decls.NewMapType(decls.String, decls.Dyn) is probably right choice then.

> 10:31 PM
You probably already figured this out, but when using Dyn, CEL support for x in y style containment checks exists so it's possible to guard map access, e.g. !(x in y) || y[x] == 'xzy'

Reply:
Apologies for not thinking this through more clearly before, but i have a couple ideas.

The simplest might be to lazily load claims  as a CEL expression accesses them by name from a map. Basically, provide CEL with a "ref.Val" (a cel runtime value) implementation that represents a map but delegates access to some code that loads a claim by name. This would require integrating more deeply into CEL but it is possible.  I would be tempted to omit this `claimsFilter` entirely if the only purpose is to prep the data available to CEL, even if you don't do this optimization for alpha.